reactos/drivers/usb/usbhub_new/pnp.c

```diff
@@ -0,0 +1,2832 @@
+    }
+
+  Exit:
+
+    USBH_CompleteIrp(Irp, Status);
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_FdoStartDevice(IN PUSBHUB_FDO_EXTENSION HubExtension,
+                    IN PIRP Irp)
+{
+    NTSTATUS Status;
+
+    DPRINT("USBH_FdoStartDevice: HubExtension - %p\n", HubExtension);
+
+    HubExtension->RootHubPdo = NULL;
+
+    Status = USBH_SyncGetRootHubPdo(HubExtension->LowerDevice,
+                                    &HubExtension->RootHubPdo,
+                                    &HubExtension->RootHubPdo2);
+
+    if (NT_SUCCESS(Status))
+    {
+        if (HubExtension->RootHubPdo)
+        {
+            Status = USBH_StartHubFdoDevice(HubExtension, Irp);
+        }
+        else
+        {
+            DPRINT1("USBH_FdoStartDevice: FIXME. start ParentDevice\n");
+            DbgBreakPoint();
+        }
+    }
+    else
+    {
+        DPRINT1("USBH_FdoStartDevice: FIXME. USBH_SyncGetRootHubPdo return - %lX\n",
+                Status);
+
+        DbgBreakPoint();
+        USBH_CompleteIrp(Irp, Status);
+    }
+
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBH_FdoQueryBusRelations(IN PUSBHUB_FDO_EXTENSION HubExtension,
+                          IN PIRP Irp)
+{
+    PDEVICE_RELATIONS DeviceRelations = NULL;
+    NTSTATUS Status = STATUS_SUCCESS;
+    LIST_ENTRY GhostPdoList;
+    KIRQL OldIrql;
+    PLIST_ENTRY PdoList;
+    UCHAR NumberPorts;
+    USHORT Port;
+    USHORT GhostPort;
+    PUSBHUB_PORT_DATA PortData;
+    PDEVICE_OBJECT PdoDevice;
+    PUSBHUB_PORT_PDO_EXTENSION PdoExtension;
+    NTSTATUS NtStatus;
+    PVOID SerialNumber;
+    PVOID DeviceHandle;
+    USB_PORT_STATUS UsbPortStatus;
+    PLIST_ENTRY Entry;
+    ULONG Length;
+
+    DPRINT("USBH_FdoQueryBusRelations: HubFlags - %lX\n", HubExtension->HubFlags);
+
+    if (!(HubExtension->HubFlags & USBHUB_FDO_FLAG_DEVICE_STARTED))
+    {
+       Status = STATUS_INVALID_DEVICE_STATE;
+       goto RelationsWorker;
+    }
+
+    if (!HubExtension->HubDescriptor)
+    {
+        Status = STATUS_UNSUCCESSFUL;
+        goto RelationsWorker;
+    }
+
+    if (!(HubExtension->HubFlags & USBHUB_FDO_FLAG_DO_ENUMERATION))
+    {
+        DPRINT("USBH_FdoQueryBusRelations: Skip enumeration\n");
+        goto RelationsWorker;
+    }
+
+    InterlockedIncrement(&HubExtension->PendingRequestCount);
+
+    KeWaitForSingleObject(&HubExtension->ResetDeviceSemaphore,
+                          Executive,
+                          KernelMode,
+                          FALSE,
+                          NULL);
+
+    NumberPorts = HubExtension->HubDescriptor->bNumberOfPorts;
+    DPRINT("USBH_FdoQueryBusRelations: NumberPorts - %x\n", NumberPorts);
+
+    Length = FIELD_OFFSET(DEVICE_RELATIONS, Objects) +
+             NumberPorts * sizeof(PDEVICE_OBJECT);
+
+    if (Irp->IoStatus.Information)
+    {
+        DPRINT1("FIXME: leaking old bus relations\n");
+    }
+
+    DeviceRelations = ExAllocatePoolWithTag(NonPagedPool, Length, USB_HUB_TAG);
```

Comment:
The general rule is to prefer PagedPool when possible, and for IRP_MN_QUERY_DEVICE_RELATIONS it is explicitly documented that you should use PagedPool.
However if you're sure Windows uses NonPagedPool and you don't want to change it, feel free to keep it this way.